feature: Optimized devContainer develop- and build times

Microchesst opened the pull request:
feat(devops): Optimize Windows Development Environment with Cross-Platform Support

## Type of Change

- [x] New feature
- [ ] Bug fix
- [x] Documentation update
- [x] Refactoring
- [ ] Security patch
- [ ] UI/UX improvement

## Description

Optimizes development environment for Windows while maintaining Linux compatibility through:
- Enhanced devcontainer with volume mounting for node_modules
- Optimized Docker build with multi-stage configuration
- Improved build and development scripts
- Updated GitLab authentication settings

## Testing

Tested:
- Devcontainer startup on Windows
- Volume mounting functionality
- Build and development scripts
- GitLab authentication
- Cross-platform compatibility

## Impact

### Improvements:
- Faster development environment setup
- Better node_modules handling
- Optimized Docker builds
- Cross-platform consistency

## Checklist

- [x] Code follows project guidelines
- [x] Changes tested on Windows
- [x] Cross-platform compatibility maintained

Microchesst committed:
feat: Optimize devcontainer configuratio

Microchesst commented:
This is the Guide i couldnt find the place to place the file :(
[DOC_GUIDE_GitLab-Runner.md](https://github.com/user-attachments/files/19437948/DOC_GUIDE_GitLab-Runner.md)

Microchesst committed:
Merge branch 'feature/distributed-demo' into fix/windows-gitlab-runner-devcontainer

prasadtalasila commented:
@Microchesst When I try the devContainer, the following error appears.

```log
Running the postCreateCommand from devcontainer.json...

[85388 ms] Start: Run in container: /bin/sh -c npm config set cache /tmp/.npm --global && yarn config set cache-folder /tmp/.yarn-cache
npm error code EPERM
npm error syscall chmod
npm error path /usr/local/etc/npmrc
npm error errno -1
npm error Error: EPERM: operation not permitted, chmod '/usr/local/etc/npmrc'
npm error     at async chmod (node:internal/fs/promises:1083:10)
npm error     at async Config.save (/usr/local/lib/node_modules/npm/node_modules/@npmcli/config/lib/index.js:749:5)
npm error     at async Config.set (/usr/local/lib/node_modules/npm/lib/commands/config.js:206:5)
npm error     at async Config.exec (/usr/local/lib/node_modules/npm/lib/commands/config.js:151:9)
npm error     at async Npm.exec (/usr/local/lib/node_modules/npm/lib/npm.js:207:9)
npm error     at async module.exports (/usr/local/lib/node_modules/npm/lib/cli/entry.js:74:5) {
npm error   errno: -1,
npm error   code: 'EPERM',
npm error   syscall: 'chmod',
npm error   path: '/usr/local/etc/npmrc'
npm error }
npm error
npm error The operation was rejected by your operating system.
npm error It is likely you do not have the permissions to access this file as the current user
npm error
npm error If you believe this might be a permissions issue, please double-check the
npm error permissions of the file and its containing directories, or try running
npm error the command again as root/Administrator.
npm error A complete log of this run can be found in: /home/node/.npm/_logs/2025-03-26T14_32_39_662Z-debug-0.log
[86110 ms] postCreateCommand from devcontainer.json failed with exit code 255. Skipping any further user-provided commands.
Done. Press any key to close the terminal.
```

The error is due to the permissions changes you are making at the end of the devContainer file.

In addition, you are removing a lot of packages. Can you please list the packages added/removed in this PR?

prasadtalasila commented:
A quick suggestion. We can have [multiple devContainers](https://containers.dev/implementors/spec/#devcontainerjson). For example,

```bash
project-root/
├── .devcontainer/
│   ├── client/
│   │   ├── devcontainer.json
│   │   ├── Dockerfile
│   │   └── other-files...
│   ├── dtaas/
│   │   ├── devcontainer.json
│   │   ├── Dockerfile
│   │   └── other-files...
├── client/
├── server/
└── other-project-files...
```
The `.devcontainer/client` can contain the required configuration just to work with the client website. The existing devcontainer can be moved to `.devcontainer/dtaas`. Please use the latest layers and packages while creating the client devcontainer.config. If there is a chance to update the existing devcontainer config, please do so. Thanks.

prasadtalasila commented on client/package.json:
```diff
@@ -19,13 +19,18 @@
+    "build:prod": "GENERATE_SOURCEMAP=false DISABLE_ESLINT_PLUGIN=true npx react-scripts build",
+    "build:analyze": "ANALYZE=true yarn build:fast",
+    "clean": "npx rimraf build/ dist/ coverage/ playwright-report/ test-results/ test.svg src.svg src/util/gitlab.json",
+    "clean:all": "yarn clean && npx rimraf node_modules/",
```
you can add the `node_modules/` folder to `clean` command.

Microchesst replied:
✅

prasadtalasila commented on client/package.json:
```diff
@@ -19,13 +19,18 @@
     "analyze": "source-map-explorer 'build/static/js/*.js'",
     "build": "npx react-scripts build",
-    "clean": "npx rimraf build/ dist/ node_modules/ coverage/ playwright-report/ test-results/  test.svg src.svg src/util/gitlab.json",
+    "build:fast": "GENERATE_SOURCEMAP=false npx react-scripts build",
```
Did you observe any significant time / size differences when doing different kinds of builds?

prasadtalasila commented on client/package.json:
```diff
@@ -19,13 +19,18 @@
     "config:test": "npx shx cp config/test.js public/env.js && npx shx cp config/test.js build/env.js",
-    "develop": "npx react-scripts start",
-    "format": "prettier --ignore-path ../.gitignore --write \"**/*.{ts,tsx,css,scss,json}\"",
+    "develop": "PORT=4000 npx react-scripts start",
```
Nice change.
The `public/env.js` is being used in hot reload. Is there also a way to copy the `config/dev.js` after each compilation step?

Microchesst replied:
I'm not sure i understand this 😅

prasadtalasila commented on docker/client.dockerfile:
```diff
@@ -1,28 +1,47 @@
-# Copy the rest of the application code to the working directory
-COPY ./client/ .
+# Configure TypeScript for faster compilation
+RUN echo '{"compilerOptions": {"skipLibCheck": true}}' > ./tsconfig.dev.json
```
This Dockerfile was built under two environments.

### Ubuntu 24.04 Desktop with docker 28.0.1

Builds successfully


### Ubuntu 22.04 Server with docker 28.0.4

Gives the following error

```log
 => [build 6/7] RUN echo '{"compilerOptions": {"skipLibChe  0.3s
 => ERROR [build 7/7] RUN yarn build                       56.9s 
------                                                           
 > [build 7/7] RUN yarn build:
0.683 yarn run v1.22.22
0.728 $ npx react-scripts build
7.018 Creating an optimized production build...
14.79 One of your dependencies, babel-preset-react-app, is importing the
14.79 "@babel/plugin-proposal-private-property-in-object" package without
14.79 declaring it in its dependencies. This is currently working because
14.79 "@babel/plugin-proposal-private-property-in-object" is already in your
14.79 node_modules folder for unrelated reasons, but it may break at any time.
14.79 
14.79 babel-preset-react-app is part of the create-react-app project, which
14.79 is not maintianed anymore. It is thus unlikely that this bug will
14.79 ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to
14.79 your devDependencies to work around this error. This will make this message
14.79 go away.
14.79   
56.08 Failed to compile.
56.08 
56.08 TS7016: Could not find a declaration file for module 'react-dom/client'. '/dtaas/client/node_modules/react-dom/client.js' implicitly has an 'any' type.
56.08   Try `npm i --save-dev @types/react-dom` if it exists or add a new declaration (.d.ts) file containing `declare module 'react-dom/client';`
56.08     2 | import * as React from 'react';
56.08     3 | import { createBrowserRouter, RouterProvider } from 'react-router-dom';
56.08   > 4 | import ReactDOM from 'react-dom/client';
56.08       |                      ^^^^^^^^^^^^^^^^^^
56.08     5 | import AppProvider from 'AppProvider';
56.08     6 | import { useURLbasename } from 'util/envUtil';
56.08     7 | import routes from 'routes';
56.08 
56.08 
56.24 npm notice
56.24 npm notice New major version of npm available! 10.9.2 -> 11.2.0
56.24 npm notice Changelog: https://github.com/npm/cli/releases/tag/v11.2.0
56.24 npm notice To update run: npm install -g npm@11.2.0
56.24 npm notice
56.32 error Command failed with exit code 1.
56.32 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
------
client.dockerfile:26
--------------------
  24 |     
  25 |     # Build the React app
  26 | >>> RUN yarn build
  27 |     
  28 |     # Production stage
--------------------
ERROR: failed to solve: process "/bin/sh -c yarn build" did not complete successfully: exit code: 1
```

Microchesst replied:
i have tried to fix it and atleast moved the updated to client.built.dockerfile

prasadtalasila reviewed (commented):
@Microchesst thanks for the PR. Please see the comments.

prasadtalasila commented:
> This is the Guide i couldnt find the place to place the file :( [DOC_GUIDE_GitLab-Runner.md](https://github.com/user-attachments/files/19437948/DOC_GUIDE_GitLab-Runner.md)

please place it in `docs/admin/gitlab/runner-Windows.md`

Microchesst committed:
feat: Implementation of the multi devcontainer

Microchesst commented:
1. I have now tried to make the multi devContainter setup, and added a readme for how to use it. i have tried to test it out and it seems to be working. 

2. I have also added the changes to the 'client.built.dockerfile'

Microchesst commented:
Here is the listed packages i removed from the client development devcontainer:

apt-get packages

ghcr.io/devcontainers-extra/features/apt-get-packages:1 (curl, graphviz, htop, net-tools, powerline)
Document generation and markdown tools

ghcr.io/devcontainers-extra/features/markdownlint-cli:1
ghcr.io/devcontainers-extra/features/mkdocs:2 (with numerous plugins)
Development utilities

ghcr.io/devcontainers-extra/features/exa:1
ghcr.io/devcontainers-extra/features/nestjs-cli:2
ghcr.io/devcontainers-extra/features/pipx-package:1
ghcr.io/devcontainers-extra/features/poetry:2
ghcr.io/devcontainers-extra/features/pre-commit:2
ghcr.io/devcontainers-extra/features/tmux-apt-get:1
ghcr.io/devcontainers-extra/features/vercel-serve:1
Removed VS Code Extensions
Git tools

donjayamanne.githistory
mhutchie.git-graph
Development utilities

humao.rest-client
johnpapa.vscode-peacock

Microchesst committed:
Fix: alot of improvements still having trouble with permission

Microchesst committed:
feat: now yarn install runs for both devcontainers

Microchesst committed:
feat: Now works much better

Microchesst commented:
@prasadtalasila The devcontainer set-up is working really well on windows now still need to figure out to test it on linux. 

I don't believe the cache will be coliding with anything, the only thing is if you were to run `yarn install` or a call like that at the same time from each devcontainer. Then they would both write to node_modules at the same time.

Microchesst committed:
feat: removed docker-compose and devcontainer.json

Microchesst committed:
doc: removed the toplevel

prasadtalasila commented on .devcontainer/README.md:
```diff
@@ -0,0 +1,57 @@
+   - Select the DTaaS folder and choose either the client or dtaas configuration
+
+
+### Option 2: Docker Compose
```
please remove this part of the documentation

prasadtalasila commented on .devcontainer/dtaas/devcontainer.json:
```diff
@@ -0,0 +1,85 @@
+
+  "forwardPorts": [4000, 4001, 5000, 8000],
+
+  "postCreateCommand": "npm config set cache /tmp/.npm && yarn config set cache-folder /tmp/.yarn-cache && cp client/.npmrc ~/.npmrc && mkdir -p .vscode && cp .devcontainer/workspace-settings.json .vscode/settings.json && sudo chown -R node:node ${containerWorkspaceFolder}/client/node_modules",
```
Copying of is giving an error because we don't have `.npmrc` file. The actual error is:

```text
Running the postCreateCommand from devcontainer.json...

[133814 ms] Start: Run in container: /bin/sh -c npm config set cache /tmp/.npm && yarn config set cache-folder /tmp/.yarn-cache && cp client/.npmrc ~/.npmrc && mkdir -p .vscode && cp .devcontainer/workspace-settings.json .vscode/settings.json && sudo chown -R node:node /workspaces/DTaaS/client/node_modules
yarn config v1.22.22
success Set "cache-folder" to "/tmp/.yarn-cache".
Done in 0.03s.
cp: cannot stat 'client/.npmrc': No such file or directory
[134589 ms] postCreateCommand from devcontainer.json failed with exit code 1. Skipping any further user-provided commands.
Done. Press any key to close the terminal.
```
This comment is might be true for the other devcontainer as well

Microchesst commented:
@prasadtalasila I have just played around with the `source=` we talked about and it´s simply just a way to give the volume a name 

Top one without source, bottom one with source 
![image](https://github.com/user-attachments/assets/d957c8bd-d69b-404d-a8f2-263d9348e8df)

Microchesst committed:
revert: changes to client.dockerfile & client.built.dockerfile

prasadtalasila commented:
```
  "postCreateCommand": "npm config set cache /tmp/.npm && yarn config set cache-folder /tmp/.yarn-cache && mkdir -p .vscode && cp .devcontainer/workspace-settings.json .vscode/settings.json && sudo chown -R node:node ${containerWorkspaceFolder}/client/node_modules",
```

Microchesst committed:
fix with ./npmrc and documentation cleanup
